### DISCUSSION TITLE
building lxqt-panel from master succeeds, but running the panel fails with missing symbol

### DISCUSSION THREAD

**Discussion by Frtzqe** [lxqt/lxqt#2636] [Q&A] [answered]
Maybe this question should be at some other place. The #lxqt-dev hashtag to http://www.oftc.net (which I found on the lxqt-project.org homepage did not help me). 

Anyone tried to build (on Arch) the whole suite? 

I cloned the lxqt suite, did the submodules trick, and used the ./build_all.sh script. It compiles fine on a fresh Arch system, after adding I believe only four dependencies (gtk-doc, cmake, intltool, libgrabstat). Arch is heaven compared to building it on Debian which I succeeded with some 18 months ago (but that is another story). 

Everything seems to work, but there is a problem with **lxqt-panel**. 

`lxqt-panel: symbol lookup error: lxqt-panel: undefined symbol: _ZN4LXQt10GridLayout13setItemsOrderENS0_10ItemsOrderE
`

I traced it back - well, IMHO, I hardly have seen the code base of lxqt, so be friendly with me - to a recently added feature to the master branch: the possibility to reverse order of tray items (on lxqt-panel apparently). 

I saw two changes on 1 October 2024, one in the lxqt-panel and one in the liblxqt library. 

https://github.com/lxqt/lxqt-panel/pull/2128 https://github.com/lxqt/liblxqt/pull/353 

As it seems to me, the build_all.sh script does not link lxqt-panel with the correct liblxqt.so library (as generated, and residing in /usr/local/lib on Arch). Probably the linker uses the version already installed on Arch in /usr/lib/. 

So, is this a bug, or am I doing something wrong (perhaps explictly checkout a specific version in every git repo before building, and/or need to tweek some configuration before running build_all.sh)?

**Comment by tsujan** [comment]
> It compiles fine on a fresh Arch system

Not an answer to your last question (as I don't use the build script), but LXQt 2.1.0 will be released in early November and will be packaged by Arch quickly.

> Probably the linker uses the version already installed on Arch in /usr/lib/.

Most probably.

**Reply by Frtzqe** [reply]
Sure, I know that Arch is always quick with almost any software :-) Thanks for the hint anyway. 

And I can live with Arch providing 2.1.0, but, the fun part of any non-propriatary software is that there should be people who independently check any 'free' software, and this includes the meta software like CMake scripts and other build tool scripting.

I just want to be able to build it myself (and not always the Gentoo way) :-) 

So I guess, I will just dig deeper. If I found out what is going on, I'll let it know here.

**Reply by tsujan** [reply]
> I just want to be able to build it myself

Personally, I always make installable packages from the latest sources and install them, instead of installing from sources — even in the case of my own projects. In this way, the package manager is aware of the installation. That's a piece of cake with Arch: you prepare suitable PKGBUILDs and just use `makepkg`. Arch (AUR) also supports git LXQt.

As for the build script, I used it twice on an old computer by setting `/usr` as the prefix. LXQt was installed without problem in both cases.

Anyway, your question about `/usr/local` is valid. Hopefully a dev who's more familiar with the script will answer it.

**Reply by Frtzqe** [reply]
There you got me.... you explicitly set `/usr`. I used the script, without any tweaking. And you know, the rest of `lxqt` works fine, so what is wrong with `lxqt-panel`??  And yeah, I know about PKGBUILDS, but that is already one step ahead....  Arch is providing an implementation of getting things up and running. The next stop. I am still at the first stop. 

If a project, in this case the 'lxqt-project' website provides a script (build_all.sh), one might hope that the lot is stable, and that it just works :-) 

Thanks for your time and comments.

I don't have the time rightnow, but I will investigate it later. Any help from seasoned developers is appreciated, ofcourse.

Now I come to think of it, I think it is just a bug in the make scripts. Not noticed sofar (too often), because `liblxqt` does not change that often. The fault in the script(s) just paratasise the existing `liblxqt.so` (in the sense of using it). Without problem, because it (the pretty much stable liblxqt library) 'happens' to be already installed on the system. I guess everyone who voluntary wants to build `lxqt` also uses `lxqt` already as desktop :-P

Now Tsujan recently added a new function to the library, the flaw got discovered.

**Reply by tsujan** [reply]
> Now I come to think of it, I think it is just a bug in the make scripts.

That may be true, although I'm not an expert in the build script.

You may already know it, but a warning doesn't do any harm. You didn't want to overwrite the currently installed packages — and that's very good practice because packages should be *upgraded*, not overwritten — but

*Having different versions of the same program installed in `/usr/` and `/usr/local/` could very easily result in hassles and headaches later.*

**Reply by Frtzqe** [reply]
I am totally aware of what you are warning me about, "them" double installed executables and libraries. Don't worry. Thanks. 

Another thing, I hope that someone in the **lxqt-project**-team feels responsible. Ofcourse, I can also run ``make; make install`` (and not use the overall build_all.sh script).  

But that won't work either in my Arch situation. So, please someone help me out, before I loose hours and hours of understanding how the myriad of CMake file(s) work. 

Some 18 months ago, I managed to build the lot on Debian, but that took 7 hours (perhaps more) of iterative forensic investigating and adding many "dev" packages to Debian, before the whole thing would build. Now, it will build, but `lxqt-panel` won't run :-(

**Comment by stefonarch** [comment]
My 2 cents: I used the build script with no issues with /usr/local/ (default) twice under debian, but in the wiki was always stated that `/usr` should be set as build target.

**Reply by tsujan** [reply]
> but in the wiki was always stated that /usr should be set as build target.

Yes, it's clearly explained in https://github.com/lxqt/lxqt/wiki/Building-from-source#scripts-to-build-in-one-step:

"`LXQT_PREFIX`: Base directory relative to which LXQt is installed. Points to `/usr/local` by default and should be set to `/usr`. Technically it is corresponding with CMake variable `CMAKE_INSTALL_PREFIX`."

> I used the build script with no issues with /usr/local/ (default) twice under debian

Without having LXQt packages installed; yes?

My guess is that @Frtzqe's problem isn't limited to lxqt-panel and issues will surface if his package-based LXQt installation is removed.

**Reply by Frtzqe** [reply, marked as answer]
Thank you all for your responses. I think I get it now. How it works is not what I had expected, but _fine_. I can live with it.

**Reply by Frtzqe** [reply]
**UPDATE**
I've figured it out. I guess not many people want to build an extra marginally changed `lxqt-panel` and run it (temporarily) instead of the one in lieu provided by the currently running `lxqt` desktop.  When I double checked it - after I finished some real work on a new plugin I want to build - I might be tempted to write about the procedure. But not very soon I'm afraid.

**Reply by Frtzqe** [reply]
**UPDATE 2**
Since I am in fact only interested in running a modified version of `lxqt-panel` (for my personal development purpose), I have a short answer of running such a modified `lxqt-panel` inside a running (and stable) `pcmanfm-qt`-desktop instance. This is all about `Arch` and I do not know about other distros running LXQt. 

In essence the **why** is: my recently git cloned version of `lxqt-panel` which I will be using for adding my own plugin, uses a slightly modified version of the status notifier configuration screen (_Configure Panel_ -> _Status Notifier_ -> _Configuration_). Future releases will have an extra check box for reversing the order of the items. That function is implemented in the `liblxqt.so` library (which will be upgraded as well), and since my desktop is still running the latest released version of all of the LXQT related software, that function can not be found. This triggers the "missing symbol" which was the impetus for starting this discussion.    

**The short answer**
The `build-all.sh` script generates binaries in `/usr/local/bin` and its related libraries reside in `/usr/local/lib`. After building all successfully, it seems that **_everything_** is working fine, when LXQt is restarted, except `lxqt-panel` (which complains about the missing symbol). 

I agree with an earlier comment, I did not test everything, so perhaps not _all_ will be working. It would surprise me if it did. 

Anyway, that the panel is not started is not so hard to miss. My deduction is that LXQt executables in `/usr/local/bin` mask their official (stable) counterparts in `/usr/bin`. A listing of `ps -x` confirms that.  

So, to force my modified version of `lxqt-panel` (which as I described already uses an enhanced version of the `liblxqt.so` library) misses a symbol, and will not start. 

The dirty trick is to force the executable to use the newer version of the `liblxqt.so` library:
 $ export LD_LIBRARY_PATH=/usr/local/lib; lxqt-panel 
 
This allows me to run _my_ executable to do some development and play around with it, without possibly disabling a working panel on the desktop. 
 
This is the short answer. It is not perfect. Far from it, but it works for me (because I only need `lxqt-panel`). 
 
 **The short version of the larger answer**
Clearly, the build scripts are not meant for my style of excercise. And it will not work for some, maybe most of the other executables of LXQt. 

So, I tried to build LXQt a true pristine Arch system running XFCE  (not my LXQt system under review which I thought was pristine enough, but had already many dependencies included which are needed for using the `build_all.sh` script). Pulled in automagically thanks to some, apparently minor applications I already installed, using `yay`. It cost me a few hours to figure out which other dependencies need to be installed on Arch (besides the four that I mentioned earlier). Some ten I guess. 

And, in the end, without pulling the `lxqt` dependency in Arch as normal people do, some things still were missing. **Don't do this at home!!** Let me know if you want to know which packages need to be installed though.

I might come back with a more complete version of the larger answer, but surely that will be much later :-P
